fix(dtensor): guard against `nullptr` from `TF_TensorData` in `ExtractSmallTensorValue`

qingweizhang054 opened the pull request:
## Summary

Adds a defensive check in `ExtractSmallTensorValue` to handle cases where `TF_TensorData()` may return `nullptr` despite a successful `TFE_TensorHandleResolve()`. This prevents potential segmentation faults when working with invalid, zero-sized, or uninitialized tensors.

Also introduces a focused C++ unit test to validate the nullptr handling path. The test uses `Layout::Empty()` to ensure coverage without requiring GPU/TPU hardware, enabling rapid verification across all environments.

**Local testing:**  
`bazel test //tensorflow/dtensor/... --test_tag_filters=-gpu,-tpu`  
— All DTensor tests pass on WSL-CPU; full matrix coverage will be validated by CI.

---
## What this PR changes

| Area              | Change                                                                                                                                                                                                                                                   |
| ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Safety check**  | Add a null-pointer guard before dereferencing the raw buffer returned by `TF_TensorData(...)`; if it’s `nullptr`, set `INVALID_ARGUMENT` and return early.                                                                                               |
| **Test coverage** | New target `small_constant_optimization_test` in `tensorflow/dtensor/cc/BUILD` that asserts the function now returns `nullopt` and the correct error code when given `nullptr`. The test relies on `Layout::Empty()` to avoid any device-specific setup. |
| **Scope**         | Applies to all supported scalar types: `TF_INT32`, `TF_INT64`, `TF_FLOAT`, `TF_STRING`.                                                                                                                                                                  |
---
## Why this matters

* Prevents hard crashes when tensors are uninitialized or zero-sized.
* Strengthens DTensor constant-folding reliability across the C ↔ C++ boundary.
* Provides regression coverage to ensure the path remains safe.
---
## Risk

*Low.* The patch is a defensive check only; it does not affect behavior for valid tensors.
---
## Recommendation

Safe to merge — improves robustness and now has a self-contained unit-test to prove it.


*(Final note for reviewers: see `small_constant_optimization_test.cc` for the minimal reproduction of the null-tensor case.)*

qingweizhang054 committed:
dtensor: add regression test for ExtractSmallTensorValue(nullptr)

* Introduces `small_constant_optimization_test` to ensure
  `ExtractSmallTensorValue()`:
  - Returns `std::nullopt` when passed a null `TFE_TensorHandle`.
  - Sets `TF_INVALID_ARGUMENT` in the supplied `TF_Status`.
* Guards against future regressions after the recent null-pointer
  safety fix in small_constant_optimization.cc.

Tested:
  bazel test //tensorflow/dtensor/cc:small_constant_optimization_test  # PASSED

mihaimaruseac reviewed (changes requested):
Actually, please also add a test

mihaimaruseac commented:
We also need a test to ensure the issue is not being reintroduced, but otherwise seems good.

zqw86713 commented:
Got it — I had misunderstood earlier. I thought you wanted me to run a sanity test, but now I see you're asking me to add a regression unit test. I'll add that shortly, thanks!

zqw86713 commented:
Hi @mihaimaruseac ,

I’ve just pushed the updated commit that adds the **C++ unit-test for the `nullptr` path**. All DTensor tests pass locally with the command above. Could you take another look? Thanks!

vsytch committed:
[XLA] Make CSE customizable

PiperOrigin-RevId: 781526169

basioli-k committed:
[XLA][host_offloading] Open source host offloading executable.

PiperOrigin-RevId: 781547064

pizzud committed:
hlo_runner_agnostic_test_base: Compile both modules before executing either.

RunAndCompareTwoModules compiles and executes two modules, and if we trigger an
execution before we compile the second then we can't trace the compilations for
caching purposes.

Unfortunately RunAndCompareTwoModulesReplicated doesn't have a nice way to do
the same thing, as decomposing it needs the fundamentally-hardware-dependent
concept of the device assignment.

PiperOrigin-RevId: 781555806

allanrenucci committed:
Fix formatting in `gpu_compiler.cc`.

PiperOrigin-RevId: 781558486

jeffreyadean committed:
Added benchmark for GlobalDecreasingSizeBestFitHeap that exercises
MakeFreeChunks (indirectly by calling Alloc and Free to exercise
different #s of allocated chunks and then calls Finish()).

PiperOrigin-RevId: 781561398

ezhulenev committed:
[xla:codegen] Use FpTrunc::Name to generate implementation

PiperOrigin-RevId: 781563253

zqw86713 commented on tensorflow/dtensor/cc/BUILD:
```diff
@@ -411,3 +411,19 @@ cc_library(
 )
+
+
+cc_test(
```
Hey @mihaimaruseac ,

Thanks for the feedback! I've updated the `BUILD` file to use `tf_cc_test` instead of Bazel's native `cc_test`, and I've added the necessary `load` statement.

I appreciate you pointing that out – it's a good learning for me about TensorFlow's custom build rules.

I'm running the test locally now and will ping you again when it's done.

Let me know if there's anything else I should adjust!

mihaimaruseac commented on tensorflow/dtensor/cc/BUILD:
```diff
@@ -411,3 +411,19 @@ cc_library(
 )
+
+
+tf_cc_test(
```
You're missing the `load` statement for this, at the top of the file

zqw86713 replied:
Hey there!

Thanks for pointing that out – I actually caught it during my last round of local testing and added it in my fourth commit.

Funnily enough, before pushing that fourth commit, I noticed something interesting: without the `load("//tensorflow:tensorflow.bzl", "tf_cc_test")` line, the build failed on my local machine, but the TensorFlow CI passed without issues. I spent a little time investigating that discrepancy, which was quite a learning experience!

My latest commit addresses this, and everything should be squared away now.

jparkerh committed:
Example Extension with cpp dependency injection

Trying to come up with a cleaner way to do dependency injection inside an extension that adds minimal additional apis for those who want to implement directly at the C api. This mechanism keeps the C API cleaner for pure C implementers, but also allows for implementation at CPP for easy composability/testability etc.

PiperOrigin-RevId: 781584176

tensorflower-gardener committed:
Eliminate sdy.data_flow_edge before Shardy->HLO transformation.

PiperOrigin-RevId: 781585969

seantalts committed:
[XLA:CPU] Implement rsqrt with Newton-Raphson refinement.

Design decisions:
- rsqrt is the first and perhaps only function that uses x86-specific intrinsics.
  - We choose the intrinsics we need based on the requested floating point type and vector width.
  - This means we outsource cpu feature detection elsewhere (it will go in the VecDescs generation function).
- Only some vector width + fp types are competitive with 1 / rsqrt.
  - We will benchmark on the main architectures to see where to do the replacement
  - We will add a more holistic hlo fusion benchmark with add, mul, rsqrt to better approximate the cost of the division op.

```
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
BM_RsqrtVectorized<4, F32, kRsqrt>/process_time           310412 ns       310495 ns         2271
BM_RsqrtVectorized<4, F32, kOneOverSqrt>/process_time     222650 ns       222991 ns         3125
BM_RsqrtVectorized<8, F32, kRsqrt>/process_time           312053 ns       312342 ns         2243
BM_RsqrtVectorized<8, F32, kOneOverSqrt>/process_time     352200 ns       352706 ns         2010
BM_RsqrtVectorized<8, F64, kRsqrt>/process_time           409729 ns       409382 ns         1714
BM_RsqrtVectorized<8, F64, kOneOverSqrt>/process_time    1178363 ns      1177512 ns          582
```

PiperOrigin-RevId: 781587256

WillFroom committed:
[XLA:CPU] Add conversion from math.erf f64 to libm call

PiperOrigin-RevId: 781590068

mihaimaruseac reviewed (dismissed):
Thank you!

bmass02 committed:
Add StatType for input pipeline traces.

PiperOrigin-RevId: 781614433

tensorflower-gardener committed:
Removed an error emission for a failed legalization.

PiperOrigin-RevId: 781616159

mkuperst committed:
[XLA] Factor out the "do channel IDs carry semantic information" check.

Note that `!use_spmd_partitioning_` is likely not the right condition, but I want this to be a pure refactor first and we can adjust the condition later.

PiperOrigin-RevId: 781620792

tensorflower-gardener committed:
Set XNN_FLAG_SLINKY_NO_CHECKS in xnn_fusion_thunk.

PiperOrigin-RevId: 781621183

allanrenucci committed:
[XLA:GPU] Restrict heuristics based collective combiners to multi-host topology.

PiperOrigin-RevId: 781621893

sgarciagoogle committed:
Internal CI changes.

PiperOrigin-RevId: 781622283

ezhulenev committed:
[xla:codegen] Use FpTrunc::CreateDefinition to create xla.fptrunc definitions

PiperOrigin-RevId: 781624007

WillFroom committed:
[XLA:CPU] Add expansion of math.cbrt to pow

PiperOrigin-RevId: 781634299

tensorflower-gardener committed:
Return DeadlineExceededError when getting handler from the pool times out.

PiperOrigin-RevId: 781641050

tensorflower-gardener committed:
Turn off auto sharding stablehlo test.

PiperOrigin-RevId: 781656212

majiddadashi committed:
Fix bounds check for input offsets in quantized ops.

The inputs offsets are negative of the zero points so that should be their acceptable ranges.

PiperOrigin-RevId: 781668766

WillFroom committed:
[XLA:CPU][XLA:GPU] use correct alignment on transfer_write

PiperOrigin-RevId: 781669338

terryheo committed:
Refactor delegate_test_util to simpler to use

This change improves the usability of the test_utils::SimpleDelegate utility
by replacing the long list of boolean parameters with readable Options enum.

PiperOrigin-RevId: 781681590

bixia1 committed:
[xla:gpu] Handle buffer related custom calls.

Add executable tests.

PiperOrigin-RevId: 781699123

bixia1 committed:
[XLA] Implement operand-to-result layout propagation for custom calls

This allows propagating layout from an operand to the result, based on output-to-operand-alias information.

PiperOrigin-RevId: 781706746

mkuperst committed:
[XLA] CSE collectives with different channel IDs when it's safe to do so.

PiperOrigin-RevId: 781721931

nvgrw committed:
Reduce likelihood of collisions in precompilation using Default print options.

PiperOrigin-RevId: 781724957

ddunl committed:
Migrate uses of `XLA_TEST_BACKEND` macros to use utilities in `xla_test_backend_predicates.h`

PiperOrigin-RevId: 781737406

ddunl committed:
Move `tensorflow/third_party/llvm` to `xla/third_party/llvm`

PiperOrigin-RevId: 781763691

juliagmt-google committed:
[XLA:benchmarks] Remove comments related to testing

PiperOrigin-RevId: 781766386

tensorflower-gardener committed:
Automated Code Change

PiperOrigin-RevId: 781814264

tensorflower-gardener committed:
Reverts 1d1309b2712b001b325962ee76972f14681913da

PiperOrigin-RevId: 781828585

tensorflower-gardener committed:
Automated Code Change

PiperOrigin-RevId: 781832145

ezhulenev committed:
[xla:cpu] Remove XLA_ALIGN macro

PiperOrigin-RevId: 781843003

tensorflower-gardener committed:
Make `Execute*()` `const` for `*Executable` classes.

We cache `*Executable` objects to reuse compilation results. This means that we may call `Execute*()` on such cached objects multiple times and expect that the calls won't affect the objects' behaviors. In other words, `*Executable()` should be `const`. Marking them `const` prevents the code from accidentally mutating cached executables and causing bugs.

PiperOrigin-RevId: 781847420

tensorflower-gardener committed:
Automated Code Change

PiperOrigin-RevId: 781858106

jerryxyj committed:
No public description

PiperOrigin-RevId: 781861930

tensorflower-gardener committed:
Automated Code Change

PiperOrigin-RevId: 781875010

WillFroom committed:
[XLA:CPU] Use xla intrinsic f32 -> bf16 conversion.

PiperOrigin-RevId: 781883075

tensorflower-gardener committed:
Update GraphDef version to 2285.

PiperOrigin-RevId: 781890029

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2025-07-11

PiperOrigin-RevId: 781890727

WillFroom committed:
[XLA:CPU] Fix race condition in LowerXlaMathLibPass.

PiperOrigin-RevId: 781894153

allanrenucci committed:
[XLA:GPU] Enable heuristic based synchronous and pipelined collective combining by default.

This change only takes effect if the user does not set the `xla_gpu_{all_reduce,all_gather,reduce_scatter}_combine_threshold_bytes` flags.

PiperOrigin-RevId: 781902147

KanishAnand committed:
Use unused sharding variable

PiperOrigin-RevId: 781903248

akuegel committed:
Remove mutable buffer accessor methods from HloAliasAnalysis (NFC).

All current users don't actually need mutable access.

PiperOrigin-RevId: 781905323

WillFroom committed:
[XLA:CPU] Add log1p math lib function.

PiperOrigin-RevId: 781926524

akuegel committed:
Remove dead code from TuplePointsToAnalysis (NFC).

While there, also fix a few ClangTidy warnings.

PiperOrigin-RevId: 781947527

basioli-k committed:
Reverts 55a3ba02a6b51595dea5efaf3dce7e44ab2409c1

PiperOrigin-RevId: 781950054

tensorflower-gardener committed:
Move llvm() and llvm_setup() calls down one workspace*.bzl file.

This makes XLA consistent with TF.

PiperOrigin-RevId: 781955402

chsigg committed:
[Triton] Add triton squeeze dims pass

This pass removes size-1 dimensions from tensors to generate faster code
through Triton.

- `squeeze_dims` ops are extracted from each unit dimension of `tt.reshape`.
- A series of patterns push `squeeze_dims` up through the graph, past element-wise ops, broadcast, transpose, join, reduce.
- `squeeze_dims` are folded into `tt.load` and `tt.expand_dims`, or converted back to `tt.reshape`.

PiperOrigin-RevId: 781962093

amd-songpiao committed:
PR #28314: [ROCm] added allreduce kernel registration

Imported from GitHub PR https://github.com/openxla/xla/pull/28314

added the missing allreduce kernel registration on ROCm

it fixes all the failed AllReduceTest/AllReduceTest.* in collective_ops_e2e_test. e.g.
`bazel-bin/xla/tests/collective_ops_e2e_test_amdgpu_any --gtest_filter=AllReduceTest/AllReduceTest.AsyncAllReduce_F32_2GPUs/async_one_shot`

@xla-rotation could you review my PR, please?
Copybara import of the project:

--
6517bd88291374e9754838ac82a5d58fe4386c8d by songlin <Songlin.Piao@amd.com>:

adapt allreduce kernel registration in rocm

--
6e255aa68932a42964c048dd24ecd944a7dbf0c9 by songlin <Songlin.Piao@amd.com>:

adapt to get rid of macro usage

--
af96c9743ad6abbe323709d1721c013abb84e206 by songlin <Songlin.Piao@amd.com>:

make PutSignalFlag and WaitSignalFlag as templates with specialization for ROCm and CUDA

Merging this change closes #28314

PiperOrigin-RevId: 781992142

tensorflower-gardener committed:
Reverts 9b88203f15334e7e779bb2b3200448a07ad0245b

PiperOrigin-RevId: 781999343

metaflow committed:
[XLA:GPU] flag to control nested gemm fusion

For more granular control over generic triton emitter.

That replaces the binary `xla_gpu_unsupported_enable_generic_triton_emitter_for_gemms`.

Sample usage `--xla_gpu_unsupported_generic_triton_emitter_features=enable_nested_gemm`

Drive by update of the comment on xla_gpu_enable_llvm_module_compilation_parallelism.

PiperOrigin-RevId: 782014919

apivovarov committed:
[Autotuner] Add block level emitter backend for Triton fusion.

This change introduces the Triton block-level fusion emitter backend, which enables autotuning of tile configurations for custom Triton fusions in XLA.

This backend implements the following core interfaces:
- GetSupportedConfigs:
Enumerates all supported combinations of tile sizes for the output tensors. The generated configs can be used during autotuning to explore different performance candidates. (will be added in the next PR)
- GetDefaultConfig:
Provides a default tile configuration for a given Triton fusion, used as a fallback when no tuning data is available.
- ApplyConfig:
Applies a selected block-level fusion configuration to a Triton fusion instruction by updating its GpuBackendConfig. (will be added in the next PR)

PiperOrigin-RevId: 782032163

zacmustin committed:
Fix clang-tidy errors in `pjrt_c_api_helpers.cc`.

PiperOrigin-RevId: 782047204

seantalts committed:
[XLA:CPU] Refactor Rsqrt intrinsic with new UnaryIntrinsicBase.

PiperOrigin-RevId: 782077991

tensorflower-gardener committed:
Add traceme for IFRT executable launch

PiperOrigin-RevId: 782078942

ezhulenev committed:
[xla:cpu] Move memory allocation functions to tf2xla

PiperOrigin-RevId: 782080412

tensorflower-gardener committed:
[XLA] Refactor HloDCE for improved readability.

A minor modification is also made for dangling computations where `changed` is only set to `true` when a computation is actually removed to account for the possibility that a dangling computation might be filtered by execution_threads.

PiperOrigin-RevId: 782086875

ezhulenev committed:
[xla:codegen] Start migrating Ldexp to Intrinsic declaration

PiperOrigin-RevId: 782093202

GleasonK committed:
[SDY] Refactor MHLO dependencies where possible

PiperOrigin-RevId: 782093218

ddunl committed:
Migrate uses of `XLA_TEST_BACKEND` macros to use utilities in `xla_test_backend_predicates.h`

Reverts c9ca1c456e1032ba1e4d728fdda68bf9a46b4f92

PiperOrigin-RevId: 782094674

ddunl committed:
Now that we don't steal from `tensorflow/third_party` don't do any post hoc Copybara for TensorFlow builds

(*) Except for `tensorflow/third_party/py`

PiperOrigin-RevId: 782106216

tensorflower-gardener committed:
Reverts e4fccd6471306e8b4d4012a423376c170e6a6301

PiperOrigin-RevId: 782126042

SiqiaoWu1993 committed:
Fix sink invariant in tf.ifregion.

PiperOrigin-RevId: 782135828

sdasgup3 committed:
[Phase Compilation] Part-4: Add C++ layers to test and interact with C PJRT API.

PiperOrigin-RevId: 782147317

seantalts committed:
[XLA:CPU] Use templated intrinsic helpers for log1p and fix intrinsic name.

PiperOrigin-RevId: 782156735

nafi3000 committed:
Optimize `GlobalDecreasingSizeBestFitHeap::MakeFreeChunks` by 1.2X to 1.4X.

`BM_GlobalDecreasingSizeBestFitHeap` benchmark is optimized by up to 1.17X.

PiperOrigin-RevId: 782157151

xinxinmo committed:
[xla] Add back test targets in //xla/tools/hlo_diff on CI because the service outage

errors (https://github.com/openxla/xla/actions/runs/16007180610/job/45156461568)

have been fixed in https://github.com/openxla/xla/pull/28465.

Reverts f590575b0ed93e703ec30b29f815bf5077ce5468

PiperOrigin-RevId: 782158057

ICGog committed:
[IFRT IR] Use the auto-generate pass creation functions rather than custom defined funcs.

PiperOrigin-RevId: 782172174

felixwqp committed:
[XLA]  Extract `ExtractSplitDimSpec` from `MatchWithDynamicSlice` in collective_opt_utils.

PiperOrigin-RevId: 782183029

ezhulenev committed:
[xla:codegen] Distinguish between scalar and vector arguments/results in XLA intrinsics

PiperOrigin-RevId: 782184041

pschuh committed:
Add CommonPjRtRawBuffer::ScheduleCopyTo which is an async version of
CommonPjRtRawBuffer::CopyTo. This allows scheduling the copy async after
some dependencies have completed. Backends which have to allocate resources
to perform copies may want to override this instead to garuntee ordering.

PiperOrigin-RevId: 782185871

emilyfertig committed:
[JAX] Add an interface for cross-host DCN transfers and a PjRt-IFRT/Linux implementation

PiperOrigin-RevId: 782197772

ezhulenev committed:
[xla:codegen] Introduce Intrinsic::Type class instead of an alias

PiperOrigin-RevId: 782199950

zqw86713 commented:
Hi  @mihaimaruseac ,
Just to provide a bit of context:

I've pushed a total of 6 commits so far.

* Commits 5 and 6 were both made after running `buildifier`.
* In the 5th commit, I accidentally left out a trailing newline at the end of the file. Interestingly, that commit still passed the CI.
* I added the missing newline in the 6th commit to make it cleaner, but now the `feedback/copybara` check is failing. Unfortunately, I don’t have permission to see the error details.

Could you please help me understand what went wrong, or whether the failure is related to my latest change?

Thanks a lot.

ICGog committed:
[IFRT IR] Add pass for dumping IFRT IR graphs to dot representation

PiperOrigin-RevId: 782216238

slackito committed:
Integrate LLVM at llvm/llvm-project@f8cb7987c64d

Updates LLVM usage to match
[f8cb7987c64d](https://github.com/llvm/llvm-project/commit/f8cb7987c64d)

PiperOrigin-RevId: 782220044

tensorflower-gardener committed:
Automated Code Change

PiperOrigin-RevId: 782239761

tensorflower-gardener committed:
Update GraphDef version to 2286.

PiperOrigin-RevId: 782276736

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2025-07-12

PiperOrigin-RevId: 782276814

mihaimaruseac commented:
`feedback/copybara` errors means that there is a failure in either importing the PR into the internal system (although here `import/copybara` passed), or a failure in the internal CI. See diagram just above https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#contributor-license-agreements

In this case, the error seems to be

```
tensorflow/dtensor/cc/small_constant_optimization_test.cc:4:10: error: module //tensorflow/dtensor/cc:small_constant_optimization_test does not depend on a module exporting 'tensorflow/core/framework/node_def_util.h'
    4 | #include "tensorflow/core/framework/node_def_util.h"
      |
```

tensorflower-gardener committed:
Fix typo in xnn_fusion_thunk.cc.
Should use ifdef rather than ifndef.

PiperOrigin-RevId: 782343552

qingweizhang054 committed:
[UT] Rename test suite and remove node_def_util.h dependency

qingweizhang054 committed:
[Cleanup] Remove non-English comment to maintain consistency with TensorFlow style guide

zqw86713 commented:
@mihaimaruseac  ,  Thank you for your feedback. I'm closing this PR and have opened a new one from a clean branch, as the commit history here became messy due to multiple local revisions and squash attempts.

🔄 New PR: [https://github.com/tensorflow/tensorflow/pull/96866](https://github.com/tensorflow/tensorflow/pull/96866)

The new PR contains the same code changes, now squashed into a single, clean commit that conforms to TensorFlow’s contribution standards.

Could you please take a look when you have time? thank you!
